{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Error Handling\n",
    "\n",
    "When we code, no one guaranteed that they can't make any errors during development and production. Definitely, developers can make an error at the time of development or in production time.\n",
    "\n",
    "##### Types of errors\n",
    "1. Syntax error\n",
    "2. Exception\n",
    "\n",
    "#### Syntax error\n",
    "When people starts learning a new programming language, they might not familiar with the syntax of the programming language at a first time itself.\n",
    "\n",
    "```\n",
    "   if 10 > 0:\n",
    "       print('10 is greater')\n",
    "```\n",
    "The above syntax is correct, but sometime people used to forget to add a colon (:) and an indendation after if statement\n",
    "```\n",
    "    if 10 > 0:\n",
    "    print('10 is greater')\n",
    "```\n",
    "\n",
    "<hr>\n",
    "\n",
    "#### Exception\n",
    "Even if a statement or expression is syntactically correct, it may cause an error when an attempt is made to execute it. Errors detected during execution are called exceptions\n",
    "\n",
    "**Note:** - There is one rule widely accepted by all the people in the globe, that is *any number can't make a division by the value zero (0)*\n",
    "\n",
    "But when you try to do that things by using your knowledge in python, it may throw an exception\n",
    "```\n",
    "a = int(input('Enter a number : '))\n",
    "print(10/a)\n",
    "```\n",
    "\n",
    "When you run the above code snippets by giving the input at runtime, It will give you an output. But what happens if you give value 0 as a input\n",
    "\n",
    "Just try to execute a snippet by entering a value 0 and see what happen. You will be getting an error below\n",
    "```\n",
    "    ZeroDivisionError                         Traceback (most recent call last)\n",
    "    <ipython-input-4-5f4980495e3c> in <module>\n",
    "          1 a = int(input('Enter a number : '))\n",
    "    ----> 2 print(10/a)\n",
    "\n",
    "    ZeroDivisionError: division by zero\n",
    "```\n",
    "\n",
    "We can't give an assurance that user will not give an input 0 at a runtime. They may give or may not. But we need to make sure that our program should run completely without terminating in a middle\n",
    "\n",
    "We need to catch that exception if it occurs\n",
    "\n",
    "```\n",
    "    try:\n",
    "        a = int(input('Enter a number : '))\n",
    "        print(10/a)\n",
    "        print('This statement works if no error')    # Line 4\n",
    "    except ZeroDivisionError as z:\n",
    "        print(z)\n",
    "    except ValueError:\n",
    "        print('Value you entered is not in a format')\n",
    "    except:\n",
    "        print('Unknown error')\n",
    "```\n",
    "\n",
    "- If you entered input as 0, ZeroDivisionError block will execute without evaluating the *Line 4* and other except block won't run\n",
    "- If you entered input as z (alphabets), ValueError block will execute without evaluating the *Line 4* and other except block won't run\n",
    "- If you can't predict the error will gonna happen, you can catch any exception by simply writing *except* block at the end of try...except statement\n",
    "\n",
    "If you want to run a code blocks even when an errors or not, we should write that code inside the *finally* block\n",
    "\n",
    "```\n",
    "    try:\n",
    "        a = int(input('Enter a number : '))\n",
    "        print(10/a)\n",
    "    except ZeroDivisionError as z:\n",
    "        print(z)\n",
    "    except ValueError:\n",
    "        print('Value you entered is not in a format')\n",
    "    except:\n",
    "        print('Unknown error')\n",
    "    finally:\n",
    "        print('This statement works if no error')    # last line\n",
    "```\n",
    "\n",
    "Try executing all the above code snippets and find the difference by yourself\n",
    "\n",
    "All the exceptions made by the Python is already written by the Python Community Developers when they developed Python. We can call those exceptions are **Built-in exceptions**\n",
    "\n",
    "**Note:** - To know more about Built-in exceptions, [Click here](https://docs.python.org/3.8/library/exceptions.html#bltin-exceptions)\n",
    "\n",
    "\n",
    "#### User-defined Exception\n",
    "\n",
    "Instead of go with Built-in exception, what developer can do if they want their own type of exception. Python gives you an option to create a user-defined exception\n",
    "\n",
    "Please check the code snippets below\n",
    "\n",
    "```\n",
    "    '''\n",
    "        Create a class by inheriting a parent class Exception and define __init__ function by passing \n",
    "        one required parameter *message*\n",
    "    '''\n",
    "    class MyError(Exception):\n",
    "\n",
    "        def __init__(self, message):\n",
    "            self.message = message\n",
    "\n",
    "    '''\n",
    "        To raise an exception by the user, *raise CustomException('message')*\n",
    "    '''\n",
    "    try:\n",
    "        raise MyError('My own error')\n",
    "    except MyError as m:\n",
    "        print(m)\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
